spec/integration/application/agent_spec.rb

```diff
@@ -0,0 +1,257 @@
+    Puppet[:hostprivkey] = File.join(PuppetSpec::FIXTURE_DIR, 'ssl', '127.0.0.1-key.pem')
+    Puppet[:hostcert] = File.join(PuppetSpec::FIXTURE_DIR, 'ssl', '127.0.0.1.pem')
+
+    facts = Puppet::Node::Facts.new(Puppet[:certname])
+    Puppet::Node::Facts.indirection.save(facts)
+  end
+
+  let(:server) { PuppetSpec::Puppetserver.new }
+  let(:agent) { Puppet::Application[:agent] }
+  let(:logfile) { tmpfile('agent.log') }
+  let(:args) { ['--test', '--logdest', logfile] }
+
+  context 'server_list' do
+    it "uses the first server in the list" do
+      Puppet[:server_list] = '127.0.0.1'
+
+      server.start_server do |port|
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(0)
+      end
+    end
+
+    it "falls back, recording the first viable server in the report" do
+      server.start_server do |port|
+        Puppet[:server_list] = "puppet.example.com,#{Puppet[:server]}"
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(0)
+
+        report = Puppet::Transaction::Report.convert_from(:yaml, File.read(Puppet[:lastrunreport]))
+        expect(report.master_used).to eq("127.0.0.1:#{port}")
+      end
+    end
+
+    it "omits master_used it no servers could be contacted" do
+      # PENDING: we should still write a report and try to send it
+      Puppet[:server_list] = "puppet.example.com"
+
+      # This doesn't actually need a server, because we never send the report
+      # but we should send one.
+      server.start_server do |port|
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(1)
+      end
+
+      expect(File).to_not be_exist(Puppet[:lastrunreport])
+    end
+
+    it "omits master_used when not using server_list" do
+      server.start_server do |port|
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(0)
+      end
+
+      report = Puppet::Transaction::Report.convert_from(:yaml, File.read(Puppet[:lastrunreport]))
+      expect(report.master_used).to be_nil
+    end
+  end
+
+  context 'rich data' do
+    let(:catalog) { Puppet::Resource::Catalog.new(Puppet[:certname], 'production') }
+    let(:formatter) { Puppet::Network::FormatHandler.format(:rich_data_json) }
+
+    it "applies deferred values" do
+      catalog_handler = -> (req, res) {
+        params = {
+          "type" => "Notify",
+          "title" => "deferred",
+          "parameters" => {
+            "message" => {
+              "__ptype" => "Deferred",
+              "name" => "join",
+              "arguments" => [[1, 2, 3], ":"]
+            }
+          }
+        }
+        catalog.add_resource(Puppet::Resource.from_data_hash(params))
```

Comment:
Would be nice to use inline puppet code instead of building the catalog:

```ruby
catalog = compile_to_catalog(<<-MANIFEST, node)
  notify { 'deferred':
    message => Deferred('join', [[1,2,3], ':'])
  }
MANIFEST
```

Reply:
Done